### PR TITLE
Blog post about TCE/expert system

### PR DESCRIPTION
**Pull request by s-andringa**

### CODE REVIEW

File: _posts/2019-07-24-pattern-matching-expert-system-classification.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+> If at source X it is categorized as either milk or yoghurt then it must be dairy.
+
+### What are the benefits of an expert system?
+
+First of all, it can help narrowing down machine learning predictions. For instance, if at a generic level the expert system determines that a given product must be non-food, we can rule out all machine learning predictions for categories that are not in the same subtree of the category taxonomy, in this case food categories.
+
+Additionally, a machine learning-based system needs a growing amount of validated training data in order to improve itself. This means that its predictions will need to be manually validated and corrected where necessary, before they can be fed back to the system. Contrarily, predictions made by an expert system do not always require manual validation. Not only does this save manual labour, it can also automatically generate legitimate training data and therefore help improving machine learning predictions.
+
+Finally an expert system is not a black box like many machine learning-based systems. This makes it easier to inspect, debug and adapt.
```
Comment (wvengen) [inline _posts/2019-07-24-pattern-matching-expert-system-classification.md]: Neural networks are most known for this, others like random forest or bayesian networks are more introspectable. Maybe mention neural networks explicitely?

File: _posts/2019-07-24-pattern-matching-expert-system-classification.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+## Pattern matching
+
+Pattern matching is a technique used to match a pattern against sequences of tokens. The pattern describes what a sequence must look like in order to match. Probably the most common example of pattern matching is _regular expressions_. Using a regular expression the conditions from the above rule can be expressed as:
+
+```
+fruit.*>\s*apples?(\s*>|$)
+```
+
+There are some problems with regular expressions that make it a suboptimal choice for describing our rules. They are hard to learn, they quickly become complex and hard to read, and they are too generic and powerful. All of this increases the chance of mistakes.
```
Comment (wvengen) [inline _posts/2019-07-24-pattern-matching-expert-system-classification.md]: This is especially true for non-developers.

File: _posts/2019-07-24-pattern-matching-expert-system-classification.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+It is designed to be easy to learn for non-technical users, and easy to read and understand. In the simplest scenario an expression mimics the category it is supposed to match, e.g. "fruit > apple" can be matched with the expression `fruit > apple`.
+
+Using a TCE the above rule is expressed as:
+
+```
+>> %fruit% >> apple | apples
+```
+
+TCE features a.o. child and descendant matching, wildcard matching, and negation. Expressions match subtrees by default. Whitespace and special characters are ignored and matching is always case insensitive. Read more about these and other language features in the [TCE language reference](https://github.com/q-m/tiered_category_expressions/blob/master/LANGREF.md). Also TCEs can be transpiled to regular expressions, so the actual matching can be delegated to any platform that supports regular expressions. The Ruby implementation can be found [here](https://github.com/q-m/tiered_category_expressions).
```
Comment (wvengen) [inline _posts/2019-07-24-pattern-matching-expert-system-classification.md:80]: Is it clear what a _subtree_ is?

File: _posts/2019-07-24-pattern-matching-expert-system-classification.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+We only recently started using TCEs with our expert system, though here are some preliminary conclusions and observations.
+
+- It still takes time to write proper rules. This is mostly an up-front task. Once you have a set of good, change-resilient rules, you'll only need to check and update rules when you integrate new sources.
+- Writing rules requires a lot of precision and care. TCE features like wildcards or negation can easily undermine strict unambiguous rules.
+- Non-technical users still need time to get familiar with the TCE syntax.
+- When creating rules, having some validated data provides invaluable insight into real world categories, how they map to target categories, and what exceptions may occur.
+- With a registry of hundreds of rules it becomes hard to detect possible duplicates or overlapping rules (when two rules apply to a single category path).
+- We needed to extend our system to deal with some exceptions. E.g. sometimes a rule must only apply to a specific source and then take precedence over more generic rules.
+- Source data cannot always be trusted to be correct, you might still need some sort of manual validation or error detection in place.
```
Comment (wvengen) [inline _posts/2019-07-24-pattern-matching-expert-system-classification.md:92]: Is that something worth elaborating on? Maybe guide us through a real-world case where you conclude with validation practices?

**Review by wvengen — APPROVED** [review]
Nice!!
Some notes / requests :)

File: _posts/2019-07-24-pattern-matching-expert-system-classification.md (hunk 3 above)
Comment (s-andringa) [reply _posts/2019-07-24-pattern-matching-expert-system-classification.md:80]: Maybe not entirely, but it's not important. Readers may navigate to the lang ref for more details and examples.

File: _posts/2019-07-24-pattern-matching-expert-system-classification.md (hunk 4 above)
Comment (s-andringa) [reply _posts/2019-07-24-pattern-matching-expert-system-classification.md:92]: It's not much more than a suggestion that hasnt really been explored.

**Pull request merged**